kafka/pykafka/consumer.py

```diff
@@ -47,9 +199,63 @@ class OwnedPartition(object):
     Used to keep track of offsets and the internal message queue.
     """
 
-    def __init__(self, partition):
+    def __init__(self, partition, consumer):
         self.partition = partition
+        self.consumer = consumer
         self._messages = Queue()
+        self.last_offset_consumed = 0
+        self.next_offset = 0
+
+        if self.consumer._auto_commit_enable and self.consumer.consumer_group is not None:
+            self.last_offset_consumed = self._fetch_last_known_offset()
 
     def consume(self, timeout=None):
+        """Get a single message from this partition
+        """
+        if self._messages.empty():
+            self._fetch(timeout=timeout)
```

Comment:
This is another thing which should exist on a separate thread, but I'm going to throw a monkey wrench in the code.  Each partition shouldn't be fetching from brokers for messages.  This would create one thread per partition and a lot of simultaneous requests to the broker.  Kafka 0.8 is designed such that a single requests to the broker can ask for data from multiple topics/partitons.  https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-FetchAPI

Instead of having on thread for each OwnedPartition, the SimpleConsumer should start a fetch thread which regularly polls all owned partitions for any where `len(op.messages.qsize()) < self._min_queued_messages`, organizes them by broker, and fires off the minimum number of requests to the broker.

Note: This may contradict something I told you before.  If so, you have my apologies. I kind of forgot I wanted to do this until right now.  The thread-per-partition thing is part of why samsa is a bit on the slow side when you have a ton of partitions.